In [1]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import time

# Question 1

## Part 1
### Preprocessing

In this section we will go through the data 
Reading the data and some preprations
<br>

The training contains float data but they are in string format so we must split them
delimiter is space and the first character is space, So we will go through string data and convert it to float 


In [2]:
#################### read dataset ####################
def process_string_dataset(string_data, delimiter = ' '):
    """
    convert the data we have in string to float
    string_data can be lines of strings
    """
    ## count of data lines
    data_line_count = len(string_data)

    ## Data is our data values in float
    data_floats = []
    ## create a temp string to save each numbers character
    temp_string = ''

    ## go through the data and convert it to float
    for index in range(0, data_line_count):
        data_floats.append([])
        for character in string_data[index][1:]:
            if(character != delimiter and character != '\n'):
                temp_string += character
            else:
                ## if the string goes to delimiter convert the data into float
                ## and reset temp_string variable to save the next data
                data_floats[index].append(float(temp_string))
                temp_string = ''
                
    return data_floats



def read_dataset(name):
    """
    read our given dataset and return the datas
    name is the dataset directory with its name can be forexample: ../toy/Atrain
    """
    file = open(name, 'r')

    lines = file.readlines()
    data_lines = []
    for line in lines:
        ## dataset with # in the start does not contain data
        if(line[0] != '#'):
            data_lines.append(line)
            
    ## convert to float and return the ready dataset
    float_data = process_string_dataset(data_lines)
    return float_data

In [3]:
train_raw_data = read_dataset('toy dataset/Atrain')
np.array(train_raw_data).shape

(4, 1000)

since the data was 2*1000 (For each class) we create two matrixes <br>
two matrixes was made and the two shows the 2 classes we have


In [4]:
## create the claases matrix as in file
train_class1 = np.matrix(train_raw_data[:2])
train_class2 = np.matrix(train_raw_data[2:])

Convert Dataset into pandas dataframe 

In [5]:
def attach_classes(data_frames, labels):
    """
    Consider each dataframe a class of our data
    attach the dataframes into one and add the class lebels
    Inputs:
    data_frames is an array of data_frames
    lebels is the labels to the classes
    """
    
    data_frame_count = len(data_frames)
    
    ## create an empty dataframe
    df = pd.DataFrame()
    
    for i in range(0, data_frame_count):
        ## create the label array for each class with the class data count
        class_label = np.full(len(data_frames[i]), labels[i])
        ## add the label to each class
        data_frames[i]['label'] = class_label 
        df = df.append(data_frames[i], ignore_index=True)
        
    return df

## convert each class to a dataframe
def convert_to_pd_dataframe(matrix_data):
    """
    convert a dataset in the form of matrix into pandas dataframe
    Inputs:
    matrix_data with two feature space
    """
    df = pd.DataFrame(matrix_data, columns=['feature_one', 'feature_two']) 
    
    return df
def convert_rowData_df(row_Datas, labels):
    """
    convert raw data into dataframes
    Input:
    row_Datas must be an array with the raw datas,
    if we had one raw data then it must be as [raw_data]
    each raw_data is for one class
    
    lebels is the labels to the data classes
    """
    ## create an array containing the dataframes 
    df_class_arrays = []
    
    ## get the count of row datas needed to be converted
    count = len(row_Datas)
    
    ## go through each row data array and create get a dataframe for each
    for i in range(0, count):
        df = convert_to_pd_dataframe(np.matrix(row_Datas[i]))
        df_class_arrays.append(df)
    
    ## attach the dataframes of each row data (each row data represent a class)
    df = attach_classes(df_class_arrays, labels)
    
    return df
        

In [6]:
df_train = convert_rowData_df([train_class1.T, train_class2.T], [0, 1])
df_train.head()

feature_one  feature_two  label
0    -0.813596     1.423864      0
1    -0.347811    -0.494301      0
2     0.188346    -0.277319      0
3     0.167306     0.179168      0
4    -0.899921     1.323836      0

### implementing KNN algorithm
As we know about the algorithm it uses the k nearest neigber to check which class it belongs 

In [7]:
## Set the k value
K = 5

In [8]:
## load test dataset
test_raw_data = read_dataset('toy dataset/Atest')
## get classes and convert to matrix representation
test_raw_class1 = np.array(test_raw_data).T[1000:]
test_raw_class1 = np.matrix(test_raw_class1)

test_raw_class2 = np.array(test_raw_data).T[:1000]
test_raw_class2 = np.matrix(test_raw_class2)

df_test = convert_rowData_df([test_raw_class1, test_raw_class2], [0, 1])

To calculate the Euclidean distance we use the formula below
$$
\begin{equation}
    d = \sqrt{(x_1 - x_2)^2 + (y_1 - y_2)^2}
\end{equation}
$$

And our dataset has just two feature and then we can expand the above formula to use as below ( For each class we have to calculate the formula) 
$$
\begin{equation}
d = \sqrt{(feature\_one_{test} - feature\_one_{class1})^2 + (feature\_two_{test} - feature\_two_{class1})^2}
\end{equation}
$$
$$
\begin{equation}
d = \sqrt{(feature\_one_{test} - feature\_one_{class2})^2 + (feature\_two_{test} - feature\_two_{class2})^2}
\end{equation}
$$
*Note:* the equations needed to be calculated for each point. <br>
*Note 2:* The Pythagorean theorem is the more generalized of the equation in above. 

In [9]:
## this is same as above

## this variable is made to save the result of distances
distances = []

## start the test phase
for i in range(0, 2000):
    d = (df_train['feature_one'] - df_test.iloc[i]['feature_one']) ** 2
    d += (df_train['feature_two']- df_test.iloc[i]['feature_two']) ** 2
    
    d = np.sqrt(d)
    distances.append(d)

In [10]:
test_label_votes = []
for i in range(0,2000):
    ## get the nearest K points
    d = distances[i].sort_values()[:K]
    nearest_indexes = d.index
    
    labels = df_train.iloc[nearest_indexes].label
    
    ## use the mean to calculate the average vote
    vote = labels.mean()
    
    test_label_votes.append(vote)


In [11]:
## the threshold that uses to classify test data
THRESHOLD = 0.5
test_labels = np.array(test_label_votes) > 0.5
test_labels = [1 if label else 0 for label in test_labels]
classify = (test_labels == df_test.label).values

In [15]:
classify.sum()

107

we can see that 107 of test data are correctly classified with $K = 5$ <br>
Now we will capsulate the calculations in a function

In [65]:
def reform_data(data, test_data ,THRESHOLD):
    """
    Reform an array of floats between one and zero with the threshold value into integers as 0 and 1
    Inputs:
    THRESHOLD is the value to set the data whether into 0 or 1 
    data is the float array
    """
    test_labels = np.array(data) >= 0.5
    test_labels = [1 if label else 0 for label in test_labels]
    classify = (test_labels == test_data.label).values
    
    return classify

def KNN(K, train_data, test_data, features, threshold = 0.5):
    """
    The K nearest neighbor algorithm
    
    Inputs:
    K is the KNN algorithm parameter (a positive integer value)
    train_data and test_data in the format of pandas dataframe, the algorithm's food!
    features are the feature for each data
    threshold is the value to decide the result of KNN algorithm
    """
        
    ## get the algorithm start time
    start_time = time.time_ns()
    
    assert K > 0, 'K must be a positive value and also not zero!'

    test_label_votes = []
    
    ## start the test phase
    for index in range(0, len(test_data)):
        
        d = 0
        for feature in features:
            d += (train_data[feature]- test_data.iloc[index][feature]) ** 2
        d = np.sqrt(d)

        ## the index of K nearest neighbor points
        nearest_indexes = d.sort_values()[:K].index
                
        ## get the labels of nearest data
        labels = train_data.iloc[nearest_indexes].label
        
        ## use the mean to calculate the average vote
        vote = labels.mean()
        test_label_votes.append(vote)
        
    ## decide the voting process in KNN
    classified_data = reform_data(test_label_votes, test_data, threshold)

    finish_time = time.time_ns()
    print('Finished! algorithm time: ', ((finish_time - start_time) / 1000), ' miliseconds')
        
    return classified_data


In [14]:
features = np.array(df_test.columns[df_test.columns != 'label'])

check the memory and time usage of the algorithm

In [26]:
%memit test_labels = KNN(5 , df_train, df_test, features)

Finished! algorithm time:  2107474.459  miliseconds
peak memory: 149.11 MiB, increment: 0.00 MiB


As you can see the increment was about 0 mega Bytes, This means that it does not really consume alot of memory

In [27]:
test_labels.sum()

107

#### Now its time to check different K values

In [28]:
## remove the target feature
features = np.array(df_test.columns[df_test.columns != 'label'])

## save the result of different K values
labels_classified = {}
for K in [1, 2, 5, 10, 15, 20, 50, 100, 500, 1000]:
    label = KNN(K, df_train, df_test, features)
    labels_classified[str(K)] = label

Finished! algorithm time:  2160221.174  miliseconds
Finished! algorithm time:  2126278.027  miliseconds
Finished! algorithm time:  2140965.288  miliseconds
Finished! algorithm time:  2090854.29  miliseconds
Finished! algorithm time:  2080330.263  miliseconds
Finished! algorithm time:  2075208.019  miliseconds
Finished! algorithm time:  2089939.41  miliseconds
Finished! algorithm time:  2074039.985  miliseconds
Finished! algorithm time:  2084358.741  miliseconds
Finished! algorithm time:  2093722.499  miliseconds


In [29]:
for key, item in labels_classified.items():
    print('K=%s ' % key, 'True classified %s' % item.sum(), '\taccuracy: %f' % (item.sum()*100 / len(item)))


K=1  True classified 153 	accuracy: 7.650000
K=2  True classified 186 	accuracy: 9.300000
K=5  True classified 107 	accuracy: 5.350000
K=10  True classified 91 	accuracy: 4.550000
K=15  True classified 90 	accuracy: 4.500000
K=20  True classified 91 	accuracy: 4.550000
K=50  True classified 92 	accuracy: 4.600000
K=100  True classified 94 	accuracy: 4.700000
K=500  True classified 104 	accuracy: 5.200000
K=1000  True classified 123 	accuracy: 6.150000


## Part 2

Use the *Itrain* dataset <br>
split it for into 100 for validation set and 900 for training  <br>
*NOTE: we have processed the data into csv format using convert_to_csv.py script and saved them into processed_dataset directory*

In [69]:
Itrain_df = pd.read_csv('processed_dataset/Itrain.csv')

features = Itrain_df.columns[Itrain_df.columns != 'label']

history = {}
for K in [1, 2, 3, 5, 7, 9, 10]:
    ## use 900 of each class for training    
    df_training = Itrain_df[Itrain_df.label == 0][:900]
    df_training = df_training.append(Itrain_df[Itrain_df.label == 1][:900])
    
    ## now we have a dataframe with 1800 values but the indexes are not set properly
    ## Since we are directly working with index in our algorithm, we need to reset it
    df_training = df_training.reset_index(drop=True)

    ## use 100 of each class for validation
    df_validation = Itrain_df[Itrain_df.label == 0][900:]
    df_validation = df_validation.append(Itrain_df[Itrain_df.label == 1][900:])
    
    ## now we have a dataframe with 200 values but the indexes are not set properly
    ## Since we are directly working with index in our algorithm, we need to reset it
    df_validation = df_validation.reset_index(drop=True)
    
    label = KNN(K, df_training, df_validation, features)
    
    print('K=%s ' % K, 'True classified count %s from %s' % (label.sum(), len(label)), '\taccuracy: %f' % (label.sum()*100 / len(label)))    
    
    
    

Finished! algorithm time:  414542.223  miliseconds
K=1  True classified count 200 from 200 	accuracy: 100.000000
Finished! algorithm time:  404241.222  miliseconds
K=2  True classified count 200 from 200 	accuracy: 100.000000
Finished! algorithm time:  403420.411  miliseconds
K=3  True classified count 200 from 200 	accuracy: 100.000000
Finished! algorithm time:  402239.077  miliseconds
K=5  True classified count 200 from 200 	accuracy: 100.000000
Finished! algorithm time:  403473.398  miliseconds
K=7  True classified count 200 from 200 	accuracy: 100.000000
Finished! algorithm time:  403851.638  miliseconds
K=9  True classified count 200 from 200 	accuracy: 100.000000
Finished! algorithm time:  401086.326  miliseconds
K=10  True classified count 200 from 200 	accuracy: 100.000000
